app/src/main/java/com/cronus/zdone/api/TasksRepositoryImpl.kt

```diff
@@ -1,78 +1,104 @@
 package com.cronus.zdone.api
 
 import com.cronus.zdone.AppExecutors
-import com.cronus.zdone.WorkTimeManager
 import com.cronus.zdone.api.model.*
 import com.cronus.zdone.home.HomeScreen
 import io.reactivex.Observable
+import io.reactivex.ObservableEmitter
+import java.util.*
 import javax.inject.Inject
 import javax.inject.Singleton
+import kotlin.collections.HashSet
 
 @Singleton
 class TasksRepositoryImpl @Inject constructor(
-        var appExecutors: AppExecutors,
-        var zdoneService: ZdoneService,
-        var workTimeManager: WorkTimeManager
+    var appExecutors: AppExecutors,
+    var zdoneService: ZdoneService
 ) : TasksRepository {
 
-    private val CACHE_REFRESH_TIME = 2 * 60 * 1_000_000_000L
+    private val CACHE_REFRESH_TIME = 2 * 60 * 1_000_000_000L // 2 mins in nanoseconds
 
-    var cachedData: Observable<Tasks>? = null
+    var cachedData: Tasks? = null
+    var requestId: String? = null
     var lastRequestTime: Long = 0
-    var lastWorkTime = -1
+    var observers = Collections.synchronizedSet(HashSet<ObservableEmitter<Tasks>>())
+    var dataStream = observe(Observable.create<Tasks> { observer ->
+        observers.add(observer)
+        observer.setCancellable { observers.remove(observer) }
 
-    override fun refreshTaskData(): Observable<List<Task>> {
-        cachedData = Observable.empty<Tasks>().cacheWithInitialCapacity(1)
-        return fetchData(workTimeManager.getDefaultWorkTime()) { response -> response.tasksToDo }
+        cachedData?.let {
+            val timeElapsed = System.nanoTime() - lastRequestTime
+            if (timeElapsed < CACHE_REFRESH_TIME) {
+                observer.onNext(it)
+            }
+        } ?: refreshTaskData()
+    })
+
+    override fun refreshTaskData() {
+        fetchData { response ->
+            synchronized(observers) {
```

Comment:
need to manually `synchronize` here bc we are iterating over set of observers that could be modified from another thread (iteration code runs in background, add/remove observers happens on main thread).

I think technically, the synchronization of `Collections.synchronizedSet()` would handle synchronization internally, but this is more appropriate, as the documentation does not specify the exact behavior and says you should always synchronize when iterating over the set